docs/extend/api.md

```diff
@@ -12,348 +12,967 @@ To use the built-in REST API as part of an integration, see [Consuming the REST
 
-#### Pagination
+## Endpoints
+
+There is a range of methods you can use to customize the behavior of your API endpoints. We will try to go through them in this section.
 
-You can allow the number of resources being **listed** to be customized by specifying the `limit` and `maxLimit` properties on your controller:
+### Authorization
+
+You can use a callback to determine whether an actor can access an endpoint. This is done through the `visible` method on the endpoint:
 
 ```php
-    // The number of records included by default.
-    public $limit = 20;
-    
-    // The maximum number of records that can be requested.
-    public $maxLimit = 50;
+use Flarum\Api\Context;
+use Flarum\Api\Endpoint;
+
+public function endpoints(): array
+{
+    return [
+        Endpoint\Show::make()
+            ->visible(fn (Label $label, Context $context) => $context->getActor()->can('view', $label)),
+    ];
+}
 ```
 
-You can then extract pagination information from the request using the `extractLimit` and `extractOffset` methods:
+Flarum adds a couple of useful methods. The `can`, `authenticated` & `admin` methods. `can` is just the equivalent of the above example. `authenticated` checks that the actor is logged in (not a guest). `admin` checks that the actor is an admin.
 
 ```php
-$limit = $this->extractLimit($request);
-$offset = $this->extractOffset($request);
+use Flarum\Api\Endpoint;
 
-return Tag::skip($offset)->take($limit);
+public function endpoints(): array
+{
+    return [
+        Endpoint\Show::make()
+            ->authenticated()
+            ->can('view'), // equivalent to $actor->can('view', $label)
+        Endpoint\Create::make()
+            ->authenticated()
+            ->can('createLabel'), // equivalent to $actor->can('createLabel'),
+        Endpoint\Update::make()
+            ->admin(), // equivalent to $actor->isAdmin()
+    ];
+}
 ```
 
-To add pagination links to the JSON:API document, use the `Document::addPaginationLinks` method.
+### Including relationship by default
 
-#### Sorting
+We do not recommend including relationships by default. If possible, it is better to extend the specific request payloads to be made on the frontend side and add the includes there as that keeps the API responses optimized. However, if you *really* need to include a relationship by default, you can do so through the `defaultInclude` method:
+
+```php
+use Flarum\Api\Endpoint;
 
-You can allow the sort order of resources being **listed** to be customized by specifying the `sort` and `sortField` properties on your controller:
+public function endpoints(): array
+{
+    return [
+        Endpoint\Index::make()
+            ->defaultInclude(['parent']),
+    ];
+}
+```
+
+### Lifecycle Hooks
+
+Some methods on the endpoint allow you to hook certain logic into the lifecycle of the endpoint. These are `before`, `after`, and `beforeSerialization` which is often not very different from `after` but is called before it when available on the endpoint.
```

Comment:
Might be worth talking about what you might want to do here, and why you wouldn't do it in the database resource hooks.